### PR TITLE
HPCC-33154 Refactor WebAssembly Support

### PR DESCRIPTION
**Pull request by GordonSmith**
Added list / set support

## Type of change:
- [ ] This change is a bug fix (non-breaking change which fixes an issue).
- [ ] This change is a new feature (non-breaking change which adds functionality).
- [x] This change improves the code (refactor or other change that does not change the functionality)
- [ ] This change fixes warnings (the fix does not alter the functionality or the generated code)
- [ ] This change is a breaking change (fix or feature that will cause existing behavior to change).
- [ ] This change alters the query API (existing queries will have to be recompiled)

## Checklist:
- [ ] My code follows the code style of this project.
  - [ ] My code does not create any new warnings from compiler, build system, or lint.
- [x] The commit message is properly formatted and free of typos.
  - [ ] The commit message title makes sense in a changelog, by itself.
  - [x] The commit is signed.
- [ ] My change requires a change to the documentation.
  - [ ] I have updated the documentation accordingly, or...
  - [ ] I have created a JIRA ticket to update the documentation.
  - [ ] Any new interfaces or exported functions are appropriately commented.
- [x] I have read the CONTRIBUTORS document.
- [ ] The change has been fully tested:
  - [ ] I have added tests to cover my changes.
  - [ ] All new and existing tests passed.
  - [ ] I have checked that this change does not introduce memory leaks.
  - [ ] I have used Valgrind or similar tools to check for potential issues.
- [x] I have given due consideration to all of the following potential concerns:
  - [ ] Scalability
  - [ ] Performance
  - [ ] Security
  - [ ] Thread-safety
  - [ ] Cloud-compatibility
  - [ ] Premature optimization
  - [ ] Existing deployed queries will not be broken
  - [ ] This change fixes the problem, not just the symptom
  - [x] The target branch of this pull request is appropriate for such a change.
- [ ] There are no similar instances of the same problem that should be addressed
  - [ ] I have addressed them here
  - [ ] I have raised JIRA issues to address them separately
- [ ] This is a user interface / front-end modification
  - [ ] I have tested my changes in multiple modern browsers
  - [ ] The component(s) render as expected

## Smoketest:
- [ ] Send notifications about my Pull Request position in Smoketest queue.
- [ ] Test my draft Pull Request.

### CODE REVIEW

**Review by ghalliday — COMMENTED** [review]
A few comments.  Some may be invalid as I gradually understood the code better.

My main thought is that I'm not sure the string functions are correct - latin1 v utf8 v utf16.  It wasn't clear what encoding were in use, where, and some of the tests against 1<<8 looked wrong.   It needs test cases for i) ascii ii) latin1, iii) extended unicode.

File: plugins/wasmembed/component-model/src/list.hpp [hunk 1]
```diff
@@ -0,0 +1,133 @@
...
+            auto [begin, length] = store_into_range(cx, list);
+            integer::store(cx, begin, ptr);
+            integer::store(cx, length, ptr + 4);
+        }
+
+        template <typename T>
+        WasmValVector lower_flat(CallContext &cx, const list_t<T> &v)
+        {
+            std::size_t maybe_length = 0;
+            if (maybe_length)
```
[the remark this answers is not in the document]
Comment (GordonSmith) [reply plugins/wasmembed/component-model/src/list.hpp:65]: FYI - looking at the ref, it is for future support of fixed length arrays.

File: plugins/wasmembed/component-model/src/list.hpp [hunk 2]
```diff
@@ -0,0 +1,133 @@
...
+    template <List T>
+    inline list_t<typename ValTrait<T>::inner_type> load(const CallContext &cx, uint32_t ptr);
+
+    namespace list
+    {
+        template <typename T>
+        std::tuple<offset, size> store_into_range(CallContext &cx, const list_t<T> &v)
+        {
+            auto elem_type = ValTrait<T>::type;
+            size_t nbytes = ValTrait<T>::size;
```
[the remark this answers is not in the document]
Comment (GordonSmith) [reply plugins/wasmembed/component-model/src/list.hpp:31]: Strings should be stored as a pair of uint32_t, (its trait size is 8) one is a ptr to var length string and the other is its length.

File: plugins/wasmembed/component-model/src/list.hpp [hunk 3]
```diff
@@ -0,0 +1,133 @@
...
+
+        template <typename T>
+        WasmValVector lower_flat(CallContext &cx, const list_t<T> &v)
+        {
+            std::size_t maybe_length = 0;
+            if (maybe_length)
+            {
+                assert(maybe_length == v.size());
+                WasmValVector flat;
+                for (auto e : v)
```
[the remark this answers is not in the document]
Comment (GordonSmith) [reply plugins/wasmembed/component-model/src/list.hpp:69]: Removed for now (fixed length array support)

File: plugins/wasmembed/component-model/src/list.hpp [hunk 4]
```diff
@@ -0,0 +1,133 @@
...
+            uint32_t begin = integer::load<uint32_t>(cx, ptr);
+            uint32_t length = integer::load<uint32_t>(cx, ptr + 4);
+            return load_from_range<T>(cx, begin, length);
+        }
+
+        template <typename T>
+        list_t<T> lift_flat(const CallContext &cx, const WasmValVectorIterator &vi)
+        {
+            std::size_t maybe_length = 0;
+            if (maybe_length)
```
[the remark this answers is not in the document]
Comment (GordonSmith) [reply plugins/wasmembed/component-model/src/list.hpp:105]: Removed for now.

File: plugins/wasmembed/component-model/src/string.cpp [hunk 5]
```diff
@@ -0,0 +1,287 @@
...
+            {
+                ptr = cx.realloc(ptr, worst_case_size, 1, encoded.second);
+                assert(ptr + encoded.second <= cx.memory.size());
+            }
+            return std::make_pair(ptr, encoded.second);
+        }
+
+        std::pair<uint32_t, uint32_t> store_utf16_to_utf8(CallContext &cx, const char8_t *src, uint32_t src_code_units)
+        {
+            uint32_t worst_case_size = src_code_units * 3;
```
[the remark this answers is not in the document]
Comment (GordonSmith) [reply plugins/wasmembed/component-model/src/string.cpp:42]: I don't think so, from the hosts point of view as long as its writing within the guest realloc memory size and within the max size of the guest memory its fair game.  It is up to the guest to alloc the requested memory block or fail.

File: plugins/wasmembed/component-model/src/string.cpp [hunk 6]
```diff
@@ -0,0 +1,287 @@
...
+        {
+            assert(src_code_units <= MAX_STRING_BYTE_LENGTH);
+            uint32_t ptr = cx.realloc(0, 0, 2, src_code_units);
+            trap_if(cx, ptr != align_to(ptr, 2));
+            trap_if(cx, ptr + src_code_units > cx.memory.size());
+            uint32_t dst_byte_length = 0;
+            for (size_t i = 0; i < src_code_units; ++i)
+            {
+                uint32_t usv = *src;
+                if (static_cast<uint32_t>(usv) < (1 << 8))
```
[the remark this answers is not in the document]
Comment (GordonSmith) [reply plugins/wasmembed/component-model/src/string.cpp:80]: I agree - this is an odd one as the reference implementation included an "optimization" where it "hopes" the string can be stored as single byte charset, but if it hits a char which can't it switches to store the entire string as 2 bytes.

File: plugins/wasmembed/component-model/src/string.cpp [hunk 7]
```diff
@@ -0,0 +1,287 @@
...
+            }
+            uint32_t latin1_size = static_cast<uint32_t>(encoded.second / 2);
+            for (uint32_t i = 0; i < latin1_size; ++i)
+                cx.memory[ptr + i] = cx.memory[ptr + 2 * i];
+            ptr = cx.realloc(ptr, src_byte_length, 1, latin1_size);
+            trap_if(cx, ptr + latin1_size > cx.memory.size());
+            return std::make_pair(ptr, latin1_size);
+        }
+
+        std::pair<offset, bytes> store_into_range(CallContext &cx, const string_t &v)
```
[the remark this answers is not in the document]
Comment (GordonSmith) [reply plugins/wasmembed/component-model/src/string.cpp:148]: FYI string_t is 
```
    struct string_t
    {
        Encoding encoding;
        const char8_t *ptr;
        size_t byte_len;
    };
```

In our use case all strings and guests are utf8, but I should add additional Unicode testing to the unit tests.

File: plugins/wasmembed/component-model/test/host-util.cpp [hunk 8]
```diff
@@ -0,0 +1,60 @@
...
+{
+    switch (from_encoding)
+    {
+    case Encoding::Latin1:
+    case Encoding::Utf8:
+        switch (to_encoding)
+        {
+        case Encoding::Latin1:
+        case Encoding::Utf8:
+            std::memcpy(dest, src, byte_len);
```
[the remark this answers is not in the document]
Comment (GordonSmith) [reply plugins/wasmembed/component-model/test/host-util.cpp:26]: This is just a "fake host" for the unit tests, so was only concerned with utf8 <-> utf8 conversion, but I can either flesh this out (with an off the shelf lib) or fail when not utf8 (for now).

File: plugins/wasmembed/component-model/test/host-util.cpp [hunk 9]
```diff
@@ -0,0 +1,60 @@
...
+            throw std::runtime_error("Invalid encoding");
+        }
+        break;
+    case Encoding::Utf16:
+    case Encoding::Latin1_Utf16:
+        switch (to_encoding)
+        {
+        case Encoding::Latin1:
+        case Encoding::Utf8:
+        // {
```
[the remark this answers is not in the document]
Comment (GordonSmith) [reply plugins/wasmembed/component-model/test/host-util.cpp:45]: Yes I need to pick a unicode lib for the fake host...

File: plugins/wasmembed/secure-enclave.cpp [hunk 10]
```diff
@@ -396,80 +422,88 @@ class SecureFunction : public CInterfaceOf<IEmbedFunctionContext>
...
-        std::vector<wasmtime::Val> memIdxVar;
-        int32_t memIdx;
-
-        switch (typecode)
+        case type_int:
         {
-        case type_boolean:
-            memIdxVar = wasmStore->callRealloc(wasmName, {0, 0, 1, (int32_t)numElems});
-            memIdx = memIdxVar[0].i32();
+            assert(elemSize == sizeof(int32_t));
```
[the remark this answers is not in the document]
Comment (GordonSmith) [reply plugins/wasmembed/secure-enclave.cpp:453]: Yes wasm + 64bit addressable memory is being rolled out currently, but our version of the wasmtime runtime does not support it (their latest does).

**Comment by ghalliday** [comment]
@GordonSmith I don't see a commit with the changes.

**Review by ghalliday — CHANGES_REQUESTED** [review]
No changes pushed to the branch

**Comment by GordonSmith** [comment]
Closing as code it out of date now.
